ebpf/state.go

```diff
@@ -238,46 +191,103 @@ func (ns *networkState) newClient(clientID string) bool {
 	now := time.Now()
+
 	client := ns.clients[id]
 	client.lastFetch = now
 
-	// Update send/recv bytes stats
-	for key, conn := range connByKey {
-		stats, ok := client.stats[key]
-		if !ok {
-			stats = &sentRecvStats{}
-			client.stats[key] = stats
+	conns := make([]ConnectionStats, 0)
+
+	// Closed connections
+	for key, c := range client.closedConnections {
+		if c2, ok := active[key]; ok { // This closed connection has become active again
+			// If we're seeing unexpected ordering, lets not combine these two connections.
+			if c.LastUpdateEpoch >= c2.LastUpdateEpoch {
+				ns.telemetry.unorderedConns++
+				// TODO: Should we `continue`?
+			} else {
+				c.MonotonicSentBytes += c2.MonotonicSentBytes
+				c.MonotonicRecvBytes += c2.MonotonicRecvBytes
+				c.MonotonicRetransmits += c2.MonotonicRetransmits
+			}
+
+			if _, ok := client.stats[key]; !ok {
+				if len(client.stats) >= ns.maxClientStats {
+					ns.telemetry.connDropped++
+					continue
+				}
+				client.stats[key] = &stats{}
+			}
+
+			ns.updateConnWithStats(client, key, &c, false, now)
+		} else { // Since connection is no longer active, lets just remove the stats object
+			ns.updateConnWithStats(client, key, &c, true, now)
 		}
 
-		// If we have an override for this conn for this client, aggregate the conn
-		if override, ok := client.overrideConnections[key]; ok {
-			aggregateConnAndStat(conn, &override)
+		conns = append(conns, c)
+	}
+
+	// Active connections
+	for key, c := range active {
+		if _, ok := client.closedConnections[key]; ok {
+			// If this connection was both closed and reopened, update the counters to reflect only the active connection.
+			// The monotonic counters will be the sum of all connections that cross our interval start + finish.
+			if stats, ok := client.stats[key]; ok {
+				stats.totalRetransmits = c.MonotonicRetransmits
+				stats.totalSent = c.MonotonicSentBytes
+				stats.totalRecv = c.MonotonicRecvBytes
+			}
+			continue // We processed this connection during the closed connection pass, so lets not do it again.
 		}
 
-		stats.lastSent = conn.MonotonicSentBytes - stats.totalSent
-		stats.lastRecv = conn.MonotonicRecvBytes - stats.totalRecv
-		stats.lastRetransmits = conn.MonotonicRetransmits - stats.totalRetransmits
+		if _, ok := client.stats[key]; !ok {
+			if len(client.stats) >= ns.maxClientStats {
+				ns.telemetry.connDropped++
+				continue
+			}
+			client.stats[key] = &stats{}
+		}
+
+		ns.updateConnWithStats(client, key, c, false, now)
+
+		conns = append(conns, *c)
+	}
 
-		conn.LastSentBytes = stats.lastSent
-		conn.LastRecvBytes = stats.lastRecv
-		conn.LastRetransmits = stats.lastRetransmits
+	return conns
+}
 
-		stats.totalSent = conn.MonotonicSentBytes
-		stats.totalRecv = conn.MonotonicRecvBytes
-		stats.totalRetransmits = conn.MonotonicRetransmits
+func (ns *networkState) updateConnWithStats(client *client, key string, c *ConnectionStats, shouldRemoveStats bool, now time.Time) {
```

Comment:
Do we need to check the LastUpdateEpoch in here at all?

Reply:
Nope, it's checked elsewhere. I'll likely be changing the stats object to remove the `time.Time` object and have it just use epoch times to stay consistent, but that will come in another PR